### PR TITLE
Add getTimesPosix

### PR DESCRIPTION
**Pull request by FreeSlave**
I'm not sure about the order of arguments in getTimesPosix.
Maybe you want it to follow getTimesWin-like approach when the additional argument is the first among SysTime arguments.

### CODE REVIEW

File: std/file.d [hunk 1]
```diff
@@ -1463,6 +1519,68 @@ version(Windows) @system unittest
...
+            auto diffa = accessTime2 - currTime;
+            auto diffm = modificationTime2 - currTime;
+            auto diffc = changeTime2 - currTime;
+            scope(failure) writefln("[%s] [%s] [%s] [%s] [%s] [%s] [%s]",
+                                    accessTime2, modificationTime2, changeTime2, currTime, diffa, diffm, diffc);
+
+            //There is no guarantee that the access time will be updated.
+            assert(abs(diffa) <= leeway + sleepTime);
+            assert(abs(diffm) <= leeway);
+            assert(abs(diffc) <= leeway);
```
Comment (CyberShadow) [inline std/file.d]: Are these tests meaningful with `sleepTime < leeway`?

File: std/file.d [hunk 2]
```diff
@@ -1463,6 +1519,68 @@ version(Windows) @system unittest
...
+        }
+
+        assert(accessTime1 <= accessTime2);
+        assert(modificationTime1 <= modificationTime2);
```
Comment (CyberShadow) [inline std/file.d]: Why `<=` and not `<` ? AFAIK the lowest precision of filesystem timestamps is FAT32 with 2 seconds, which should still almost always pass this test with `<`.

**Comment by CyberShadow** [comment]
Pinging @andralex as he requested that his approval is obtained for any new Phobos symbols.

Summary: this adds `getTimesPosix` to complement `getTimesWin`. Currently, there is no D-ish way to get the "changed" timestamp of a file on POSIX, other than first wrapping it in a `DirEntry` and querying that.

**Comment by wilzbach** [comment]
Why not do the following to have a unified API for all OSes?

```d
Times getTimes(R)(R name)
...
struct Times
{
    SysTime accessTime;
    SysTime modificationTime;
    version(Windows) SysTime creationTime;
    version(Posix) SysTime statusChangeTime;
}
```

This will probably be a lot less confusing than having getTimes, getTimesWin and getTimesPosix which all only marginally differ.

**Comment by CyberShadow** [comment]
That's pretty much what `DirEntry` already is.

**Comment by wilzbach** [comment]
Cool! So we have a successful precedent?

**Comment by CyberShadow** [comment]
Not sure what you mean by that :)

I guess we *could* go in the other direction and instead of adding `getTimesPosix` to complement `getTimesWin`, deprecate `getTimesWin` in favor of `DirEntry`. Is that what you meant?

**Comment by CyberShadow** [comment]
BTW, perhaps that could also be a good direction for a solution for #5755? Add a `DirEntry` constructor that takes a name and `stat_t`.

**Comment by wilzbach** [comment]
Oh sorry for not being so specific. I was dumply thinking that DirEntry only works for directories, but yeah of course it does for files too.
So I was actually thinking about adding a `getTimes` function that only takes a name and returns a DirEntry-like object (or maybe even a DirEntry) and yep then we could deprecate getTimesWin which would be an additional win in my opinion.
Anyhow if we decide to add an additional constructor overload to DirEntry, we probably want to add the convenience `dirEntry`  too as that's a common pattern throughout Phobos for nice UFCS usage.

**Comment by CyberShadow** [comment]
> So I was actually thinking about adding a getTimes function that only takes a name and returns a DirEntry-like object (or maybe even a DirEntry)

My point was that would be entirely redundant because you can use `DirEntry` itself in exactly the same way:

```d
auto de = DirEntry(fileName);
writeln(de.timeLastModified);
writeln(de.timeLastAccessed);
```
etc.


> Anyhow if we decide to add an additional constructor overload to DirEntry, we probably want to add the convenience dirEntry too as that's a common pattern throughout Phobos for nice UFCS usage.

Funny of you to say that, as there *was* a `dirEntry` function not too long ago, which got deprecated in favor of the constructor.

In any case, constructors are UFCS-able, so...

```d
	string fn;
	auto de = fn.DirEntry;
```

File: std/file.d (hunk 1 above)
Comment (FreeSlave) [reply std/file.d]: I don't know, I just copied tests from getTimes.

File: std/file.d (hunk 2 above)
Comment (FreeSlave) [reply std/file.d]: Again, it's just copy-paste from getTimes. Not very good, but getTimesWin also has this copy-paste

**Comment by FreeSlave** [comment]
If we deprecate getTimesWin in favor of DirEntry, we can deprecate getTimes too.

**Comment by FreeSlave** [comment]
> BTW, perhaps that could also be a good direction for a solution for #5755? Add a DirEntry constructor that takes a name and stat_t.

Such constructor can be added, but I hardly see it as a solution. The user may want to avoid using DirEntry.

**Comment by CyberShadow** [comment]
> The user may want to avoid using DirEntry.

Why though?

**Comment by FreeSlave** [comment]
> Why though?

On Posix It's not nothrow, not pure, etc.

**Comment by andralex** [comment]
All - please educate me. Why not use `DirEntry` for this?

**Comment by FreeSlave** [comment]
In this case DirEntry is ok. I just thought we need a mirror function to complement getTimesWin.

I would like you to give some attention to #5755 which is also related to time functions and more important in my opinion.

**Comment by jmdavis** [comment]
All this does over `getTimes` is provide the value for `st_ctim`, which I very much doubt that very many people want, and honestly, I question the value of either `getTimes` or `getTimesWin` over `DirEntry`, but I believe that they predate being able to get a `DirEntry` for a file without calling `dirEntries`.

**Comment by andralex** [comment]
Cool, I approved #5755

**Comment by wilzbach** [comment]
Why was this merged?
I thought there was a consensus that this is the wrong direction and that we should rather deprecate getTimesWin instead of introducing new platform specific functions in a supposedly abstracting standard library.

Also this has never been approved by @andralex and for good reasons all new public symbols need to be approved by Andrei.

**Pull request merged**